Fix type sharing on loaded ScriptModules

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29826 Fix type sharing on loaded ScriptModules**
* #29825 Re-unify module and interface in ConcreteModuleType
* #29824 Split ConcreteModuleType into two types

After save/load, we lose concrete type information. So if you tried to
script something that contained a loaded ScriptModule as a submodule,
the following sequence happened:
1. During ConcreteType inference, the loaded submodule got a new
inferred type.
2. But it already has a type! So there was a type mismatch.

To fix this, we should generate a ConcreteType directly from the loaded
submodule type (similar to what we do for interfaces). This makes sense
too--the ConcreteModuleType should be empty, since all the "sugaredness"
was stripped out during the save/load process.

Differential Revision: [D18575009](https://our.internmc.facebook.com/intern/diff/D18575009)

suo committed:
Fix type sharing on loaded ScriptModules

After save/load, we lose concrete type information. So if you tried to
script something that contained a loaded ScriptModule as a submodule,
the following sequence happened:
1. During ConcreteType inference, the loaded submodule got a new
inferred type.
2. But it already has a type! So there was a type mismatch.

To fix this, we should generate a ConcreteType directly from the loaded
submodule type (similar to what we do for interfaces). This makes sense
too--the ConcreteModuleType should be empty, since all the "sugaredness"
was stripped out during the save/load process.

kostmo commented:
## CircleCI build failures summary
As of commit 79f58ad:
* **2/2** failures introduced in this PR
* **2/2** recognized as flaky
    * Re-run these jobs?



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=79f58ad2522d05df89d6a85fa4836d2c2ea496f6):

| Job | Step | Log excerpt |
| --- | --- | --- |
| pytorch_linux_xenial_py3_clang5_android_ndk_r19c_x86_32_build | Set Up CI Environment After attach_workspace | `E: Failed to fetch` |
| pytorch_linux_xenial_py3_clang5_asan_build | Set Up CI Environment After attach_workspace | `E: Failed to fetch` |


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

suo committed:
Update on "Fix type sharing on loaded ScriptModules"


After save/load, we lose concrete type information. So if you tried to
script something that contained a loaded ScriptModule as a submodule,
the following sequence happened:
1. During ConcreteType inference, the loaded submodule got a new
inferred type.
2. But it already has a type! So there was a type mismatch.

To fix this, we should generate a ConcreteType directly from the loaded
submodule type (similar to what we do for interfaces). This makes sense
too--the ConcreteModuleType should be empty, since all the "sugaredness"
was stripped out during the save/load process.

suo committed:
Update on "Fix type sharing on loaded ScriptModules"


After save/load, we lose concrete type information. So if you tried to
script something that contained a loaded ScriptModule as a submodule,
the following sequence happened:
1. During ConcreteType inference, the loaded submodule got a new
inferred type.
2. But it already has a type! So there was a type mismatch.

To fix this, we should generate a ConcreteType directly from the loaded
submodule type (similar to what we do for interfaces). This makes sense
too--the ConcreteModuleType should be empty, since all the "sugaredness"
was stripped out during the save/load process.

Differential Revision: [D18575009](https://our.internmc.facebook.com/intern/diff/D18575009)

suo committed:
Update on "Fix type sharing on loaded ScriptModules"


After save/load, we lose concrete type information. So if you tried to
script something that contained a loaded ScriptModule as a submodule,
the following sequence happened:
1. During ConcreteType inference, the loaded submodule got a new
inferred type.
2. But it already has a type! So there was a type mismatch.

To fix this, we should generate a ConcreteType directly from the loaded
submodule type (similar to what we do for interfaces). This makes sense
too--the ConcreteModuleType should be empty, since all the "sugaredness"
was stripped out during the save/load process.

Differential Revision: [D18575009](https://our.internmc.facebook.com/intern/diff/D18575009)
